File: javascript/make_video.js
/**
 *  make_video.js
 *
 *  Renders the Polpot Sort elimination process as an mp4 video.
 *
 *  No npm dependencies. Frames are rasterized manually into a raw RGB24
 *  pixel buffer and piped into a system `ffmpeg` process via stdin.
 *  Requires `ffmpeg` to be installed and available on PATH.
 */

'use strict';

const { spawn } = require('child_process');

const COLOR_BG = [30, 30, 30];
const COLOR_PENDING = [207, 207, 207];
const COLOR_SURVIVOR = [58, 157, 93];
const COLOR_BASELINE = [240, 190, 40];

const WIDTH = 1000;
const HEIGHT = 550;
const MARGIN_BOTTOM = 20;
const MARGIN_TOP = 20;
const BAR_GAP_RATIO = 0.15;

function mulberry32(seed) {
  let a = seed >>> 0;
  return function () {
    a |= 0;
    a = (a + 0x6D2B79F5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function sameColor(a, b) {
  return a[0] === b[0] && a[1] === b[1] && a[2] === b[2];
}

function buildFrames(data, rng) {
  const n = data.length;
  const display = data.slice();
  const colors = new Array(n).fill(COLOR_PENDING);
  const frames = [[display.slice(), colors.slice()]];

  const baselineIdx = Math.floor(rng() * n);
  const baselineVal = data[baselineIdx];
  const survivorsIdx = [baselineIdx];
  colors[baselineIdx] = COLOR_BASELINE;
  frames.push([display.slice(), colors.slice()]);

  for (let i = 0; i < n; i++) {
    if (i === baselineIdx) continue;
    if (data[i] <= baselineVal) {
      colors[i] = COLOR_SURVIVOR;
      survivorsIdx.push(i);
    } else {
      colors[i] = COLOR_SURVIVOR;
      display[i] = 0;
    }
    frames.push([display.slice(), colors.slice()]);
  }

  for (const idx of survivorsIdx) {
    if (rng() <= 0.2) {
      display[idx] = 0;
    }
    frames.push([display.slice(), colors.slice()]);
  }

  frames.push([display.slice(), colors.slice()]);
  frames.push([display.slice(), colors.slice()]);

  return frames;
}

function drawFrame(colors, display, maxVal) {
  const buf = Buffer.alloc(WIDTH * HEIGHT * 3);
  for (let p = 0; p < WIDTH * HEIGHT; p++) {
    buf[p * 3] = COLOR_BG[0];
    buf[p * 3 + 1] = COLOR_BG[1];
    buf[p * 3 + 2] = COLOR_BG[2];
  }

  const n = display.length;
  const chartH = HEIGHT - MARGIN_TOP - MARGIN_BOTTOM;
  const slotW = WIDTH / n;
  const barW = Math.max(slotW * (1 - BAR_GAP_RATIO), 1);
  const baselineBarW = Math.max(barW, 4);

  for (let i = 0; i < n; i++) {
    const val = display[i];
    if (val <= 0) continue;

    const color = colors[i];
    const w = sameColor(color, COLOR_BASELINE) ? baselineBarW : barW;
    const barH = (val / maxVal) * chartH;

    const x0 = Math.round(i * slotW + (slotW - w) / 2);
    const x1 = Math.round(x0 + w);
    const y1 = HEIGHT - MARGIN_BOTTOM;
    const y0 = Math.round(y1 - barH);

    for (let y = y0; y < y1; y++) {
      if (y < 0 || y >= HEIGHT) continue;
      const rowOffset = y * WIDTH;
      for (let x = Math.max(x0, 0); x < Math.min(x1, WIDTH); x++) {
        const p = (rowOffset + x) * 3;
        buf[p] = color[0];
        buf[p + 1] = color[1];
        buf[p + 2] = color[2];
      }
    }
  }

  return buf;
}

function makeVideo(data, outputPath, {
  seed = 42,
  targetDuration = 6.0,
  fpsMin = 10,
  fpsMax = 240,
} = {}) {
  return new Promise((resolve, reject) => {
    const rng = mulberry32(seed);
    const frames = buildFrames(data, rng);
    const maxVal = Math.max(...data);

    let fps = frames.length / targetDuration;
    fps = Math.max(fpsMin, Math.min(fpsMax, fps));

    const ffmpeg = spawn('ffmpeg', [
      '-y',
      '-f', 'rawvideo',
      '-pixel_format', 'rgb24',
      '-video_size', `${WIDTH}x${HEIGHT}`,
      '-framerate', String(fps),
      '-i', '-',
      '-c:v', 'libx264',
      '-pix_fmt', 'yuv420p',
      outputPath,
    ]);

    ffmpeg.stderr.on('data', () => {}); // suppress ffmpeg logs
    ffmpeg.on('error', reject);
    ffmpeg.on('close', (code) => {
      if (code === 0) {
        console.log(`saved: ${outputPath} (${frames.length} frames, fps=${fps.toFixed(1)}, ${(frames.length / fps).toFixed(2)}s)`);
        resolve();
      } else {
        reject(new Error(`ffmpeg exited with code ${code}`));
      }
    });

    for (const [display, colors] of frames) {
      ffmpeg.stdin.write(drawFrame(colors, display, maxVal));
    }
    ffmpeg.stdin.end();
  });
}

if (require.main === module) {
  const rng = mulberry32(1);
  const data = Array.from({ length: 300 }, () => 1 + Math.floor(rng() * 4999));

  makeVideo(data, 'polpot_sort.mp4', { seed: 42, targetDuration: 6.0 }).catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

module.exports = { PolpotSortVideo: { buildFrames, drawFrame, makeVideo } };

File: python/make_video.py
import random
import numpy as np
from PIL import Image, ImageDraw
import imageio.v2 as imageio


COLOR_BG = (30, 30, 30)
COLOR_PENDING = (207, 207, 207)
COLOR_SURVIVOR = (58, 157, 93)
COLOR_BASELINE = (240, 190, 40)

WIDTH, HEIGHT = 1000, 550
MARGIN_BOTTOM = 20
MARGIN_TOP = 20
BAR_GAP_RATIO = 0.15


def build_frames(data, seed=42):
    rng = random.Random(seed)
    n = len(data)
    display = list(data)
    colors = [COLOR_PENDING] * n
    frames = [(display.copy(), colors.copy())]

    baseline_idx = rng.randrange(n)
    baseline_val = data[baseline_idx]
    survivors_idx = [baseline_idx]
    colors[baseline_idx] = COLOR_BASELINE
    frames.append((display.copy(), colors.copy()))

    for i in range(n):
        if i == baseline_idx:
            continue
        if data[i] <= baseline_val:
            colors[i] = COLOR_SURVIVOR
            survivors_idx.append(i)
        else:
            colors[i] = COLOR_SURVIVOR
            display[i] = 0
        frames.append((display.copy(), colors.copy()))

    for idx in survivors_idx:
        if rng.random() <= 0.2:
            display[idx] = 0
        frames.append((display.copy(), colors.copy()))

    frames.append((display.copy(), colors.copy()))
    frames.append((display.copy(), colors.copy()))

    return frames


def draw_frame(colors, display, max_val):
    img = Image.new("RGB", (WIDTH, HEIGHT), COLOR_BG)
    draw = ImageDraw.Draw(img)

    n = len(display)
    chart_h = HEIGHT - MARGIN_TOP - MARGIN_BOTTOM
    slot_w = WIDTH / n
    bar_w = max(slot_w * (1 - BAR_GAP_RATIO), 1)
    baseline_bar_w = max(bar_w, 4)

    for i, (val, color) in enumerate(zip(display, colors)):
        if val <= 0:
            continue
        bar_h = (val / max_val) * chart_h
        w = baseline_bar_w if color == COLOR_BASELINE else bar_w
        x0 = i * slot_w + (slot_w - w) / 2
        x1 = x0 + w
        y1 = HEIGHT - MARGIN_BOTTOM
        y0 = y1 - bar_h
        draw.rectangle([x0, y0, x1, y1], fill=color)

    return img


def make_video(data, output_path, seed=42, target_duration=6.0,
               fps_min=10, fps_max=240):
    frames = build_frames(data, seed=seed)
    max_val = max(data)

    fps = len(frames) / target_duration
    fps = max(fps_min, min(fps_max, fps))

    writer = imageio.get_writer(output_path, fps=fps, codec="libx264",
                                 quality=10, macro_block_size=1)

    for display, colors in frames:
        img = draw_frame(colors, display, max_val)
        writer.append_data(np.asarray(img))

    writer.close()
    print(f"saved: {output_path} ({len(frames)} frames, fps={fps:.1f}, "
          f"{len(frames)/fps:.2f}s)")


if __name__ == "__main__":
    data_rng = random.Random(1)
    data = data_rng.sample(range(1, 5000), 300)
    make_video(data, output_path="polpot_sort_pillow.mp4", seed=42, target_duration=6.0)
